std/experimental/normint.d

```diff
@@ -0,0 +1,571 @@
+static import std.algorithm;
+
+import std.experimental.color : FloatTypeFor;
+
+@safe pure nothrow @nogc:
+
+
+/**
+Normalized integers express a fractional range of values within an integer data type.
+
+Unsigned integers map the values 0 -> I.max to the fractional values 0.0 -> 1.0 in equal increments.
+Signed integers represent the values -I.max -> 0 -> I.max to fractional values -1.0 -> 0.0 -> 1.0 in equal increments. I.min is outside the nominal integer range and is clamped to represent -1.0.
+
+Params: I = Any builtin integer type.
+*/
+struct NormalizedInt(I) if(isIntegral!I)
+{
+@safe:
+
+    string toString() const
+    {
+        import std.conv : to;
+        return to!string(cast(float)this);
+    }
+    unittest
+    {
+        import std.conv : to;
+        assert(to!string(NormalizedInt!ubyte(0xFF)) == "1");
+    }
+
+pure nothrow @nogc:
+
+    I value;
+
+    /** Integral storage type. */
+    alias IntType = I;
+
+    /** Maximum integral value. */
+    enum max = I.max;
+    /** Minimum integral value. */
+    enum min = isSigned!I ? -I.max : 0;
+    /** Maximum floating point value. */
+    enum max_float = 1.0;
+    /** Minimum floating point value. */
+    enum min_float = isSigned!I ? -1.0 : 0.0;
+
+    /** Construct a NormalizedInt from an integer representation. */
+    this(I value)
+    {
+        this.value = value;
+    }
+    ///
+    unittest
+    {
+        static assert(NormalizedInt!ubyte(100) == 100);
+    }
+
+    /** Construct a NormalizedInt from a floating point representation. The value is clamped to the range [min, max]. */
+    this(F)(F value) if(isFloatingPoint!F)
+    {
+        if(value >= max_float)
+            this.value = max;
+        else if(value <= min_float)
+            this.value = min;
+        else if(value >= 0)
+            this.value = cast(I)(value*max + 0.5);
+        else
+            this.value = cast(I)(value*max - 0.5);
+    }
+    ///
+    unittest
+    {
+        static assert(NormalizedInt!ubyte(1.0) == 255);
+        static assert(NormalizedInt!ubyte(2.0) == 255);
+        static assert(NormalizedInt!ubyte(0.5) == 128);
+        static assert(NormalizedInt!ubyte(-2.0) == 0);
+        static assert(NormalizedInt!byte(-2.0) == -127);
+    }
+
+    NormalizedInt!I opUnary(string op)() const
+    {
+        return NormalizedInt!I(mixin(op ~ "value"));
+    }
+    unittest
+    {
+        static assert(+NormalizedInt!ubyte(0.5) == 128);
+        static assert(-NormalizedInt!ubyte(0.5) == 128);
+        static assert(~NormalizedInt!ubyte(0.5) == 127);
+        static assert(+NormalizedInt!byte(0.5) == 64);
+        static assert(-NormalizedInt!byte(0.5) == -64);
+        static assert(~NormalizedInt!byte(0.5) == -65);
+    }
+
+    NormalizedInt!I opBinary(string op)(NormalizedInt!I rh) const if(op == "+" || op == "-")
+    {
+        auto r = mixin("value " ~ op ~ " rh.value");
+        r = std.algorithm.min(r, max);
```

Comment:
Is this code correct for `uint` and `ulong` types? I realize these are not needed for std.color, but as this type supports all `isIntegral` types it should work (or limit the template to byte/short types). Please also add unittests for ulong if you want to support these types. (I think overflow/integer promotion semantics can lead to different results for ulong in some cases?)